core/lib/zksync_core/src/state_keeper/keeper.rs

```diff
@@ -234,6 +218,72 @@ impl ZkSyncStateKeeper {
         Err(Error::Canceled)
     }
 
+    /// This function is meant to be called only once during the state-keeper initialization.
+    /// It will check if we should load a protocol upgrade transaction, wait for it to be picked up
+    /// by the event watcher if it's a `SetChainIdUpgrade`, perform some checks and return it.
+    pub(super) async fn load_protocol_upgrade_tx(
+        &mut self,
+        pending_miniblocks: &[MiniblockExecutionData],
+        protocol_version: ProtocolVersionId,
+        l1_batch_number: L1BatchNumber,
+    ) -> Result<Option<ProtocolUpgradeTx>, Error> {
+        let first_batch_in_shared_bridge =
+            l1_batch_number == L1BatchNumber(1) && !protocol_version.is_pre_shared_bridge();
+        let previous_batch_protocol_version =
+            self.io.load_previous_batch_version_id().await.unwrap();
+
+        let version_changed = protocol_version != previous_batch_protocol_version;
+        let mut protocol_upgrade_tx = self.io.load_upgrade_tx(protocol_version).await;
+
+        // After the Shared Bridge is integrated,
+        // there has to be a setChainId upgrade transaction after the chain genesis.
+        // It has to be the first transaction of the first batch.
+        // If it's not present, we have to wait for it to be picked up by the event watcher.
+        // The setChainId upgrade does not bump the protocol version, rather attaches an upgrade
+        // transaction to the genesis protocol version.
+        while first_batch_in_shared_bridge && protocol_upgrade_tx.is_none() {
+            tracing::info!("Waiting for setChainId upgrade transaction to be picked up");
+            if self.is_canceled() {
+                return Err(Error::Canceled);
+            }
+            tokio::time::sleep(Duration::from_secs(1)).await;
+            protocol_upgrade_tx = self.io.load_upgrade_tx(protocol_version).await;
```

Comment:
To repeat the question we've discussed previously: for ENs, `load_upgrade_tx()` [always returns](https://github.com/matter-labs/zksync-era/blob/758f4877433ebdd86cc90dd592db9a7d87598586/core/lib/zksync_core/src/sync_layer/external_io.rs#L544-L545) `None`.  So it seems that the EN will enter an infinite loop if `first_batch_in_shared_bridge` is true. Is `load_protocol_upgrade_tx()` not supposed to be called for ENs? Is this implemented right now?

Reply:
This loop will be removed when [Misha's suggestion](https://github.com/matter-labs/zksync-era/pull/1012#discussion_r1481209538) is implemented, so ENs should still work.

Reply:
Hmm, it will break an assertion a few lines below though. I'll fix it